### ISSUE TITLE
Decouple modules from HTTP client (currently okhttp/retrofit)

### ISSUE THREAD

**Issue by langchain4j** [langchain4j/langchain4j#1044]
Modules should be decoupled from the HTTP client to allow using other HTTP clients than okhttp (e.g. RestClient for Spring Boot).

**Comment by kchaber** [comment]
The inspiration might be the Spring’s `RestTemplate`, which allows defining mulitple http backends through the [ClientHttpRequestFactory](https://github.com/spring-projects/spring-framework/blob/69eaf8f8289436d154823e63d71e5102a719ddac/spring-web/src/main/java/org/springframework/http/client/ClientHttpRequestFactory.java). There are multiple [implementations](https://github.com/spring-projects/spring-framework/tree/69eaf8f8289436d154823e63d71e5102a719ddac/spring-web/src/main/java/org/springframework/http/client) for common http providers, which can be passed via `RestTemplate.setRequestFactory(…)`.

**Comment by One16-KS** [comment]
Retrofit has a hard dependency on okhttp.. should we throw retrofit in the mix as well?  Retrofit could be replaced by competing libs such as openfeign?

**Comment by langchain4j** [comment]
Right, I meant okhttp+retrofit. I would avoid reflection and dynamic proxies as much as possible, I guess we can live without this convenience.

**Comment by langchain4j** [comment]
@kchaber @One16-KS could you please review the first draft: https://github.com/langchain4j/langchain4j/pull/1103 ?

**Comment by One16-KS** [comment]
Planning to do so this weekend!

**Comment by One16-KS** [comment]
Nothing to add really, really good first step, I'll try to chip in if I have spare time.

**Comment by kenix** [comment]
Hint: Spring-Web's `RestClient` is the replacement for `RestTemplate`.

As of version 0.36.2 of langchain4j a streaming chat model always waits for 1 minute to shutdown, even though we can make sure no new requests and all onCompletes are already finished. The `OkHttpClient` is completely hidden e.g. in `OllamaClient`. Is there way to gracefully shutdown the "OkHttp Dispatcher" as of now?

**Comment by langchain4j** [comment]
Unfortunately no way to shutdown yet. The plan is to migrate Ollama to Java 11 HTTP Client soon and make http client customizable. Since this is an Okhttp issue, it should be solved by this migration.

**Comment by kenix** [comment]
JDK `HttpClient` is not bad as an option (had to move away from it, due to e.g. https://bugs.openjdk.org/browse/JDK-8308364). A good comparison can be found https://www.wiremock.io/post/java-http-client-comparison.

What @kchaber suggested will reduce the effort to doctor on langchain4j due to HTTP lib details/issues in the long run (it's not langchain4j's focus anyway). Therefore it is understandable that I also vote for an abstraction similar to Spring-Web's `RestClient` or `RestTemplate` instead of depending on one specific HTTP lib.
